### ISSUE TITLE
DateTimeImmutable construction

### ISSUE THREAD

**Issue by shadowhand** [thephpleague/period#21]
I see a lot of this:

```php
new DateTimeImmutable($datetime->format('Y-m-d H:i:s.u'), $datetime->getTimeZone());
```

Wouldn't it be better to use `createFromFormat` because you always have the exact date already? For instance:

```php
DateTimeImmutable::createFromFormat(self::ISO8601, $datetime->format(self::ISO8601), $datetime->getTimeZone());
```

This reuses the format you've already defined, instead repeating `'Y-m-d H:i:s.u'` every time a new date object is made.

**Comment by nyamsprod** [comment]
hi @shadowhand,

There are two reasons as for why I'm not using `DateImmutable::createFromFormat` with the already defined constant

- To correctly use the `Period::IS88601` constant you need your `DateTimeInterface` to be in UTC as per 8601 definition. So the code should be 

```php
DateTimeImmutable::createFromFormat(
    self::ISO8601,
    $datetime->setTimeZone(new DateTimeZone('UTC'))->format(self::ISO8601),
    $datetime->getTimeZone()
);
```

And IMHO this is over-engineered!!

- The second problem is [HHVM bug](https://github.com/facebook/hhvm/issues/5152). TL;DR: in HHVM any modification made truncate/remove the microseconds info from DateImmutable. So the less modification you do the better you stay compatible with HHVM.

So bottom line, what I could do is improve/update the constants like this

```php
const DATE_LOCALE  = 'Y-m-d H:i:s.u';
const DATE_ISO8601 = 'Y-m-d\TH:i:s.u\Z';
```

Thoughts ?

**Comment by shadowhand** [comment]
Ah, I was not aware of the UTC issue. Having a constant for the format would be better than nothing.

**Comment by nyamsprod** [comment]
this issue has been addressed and resolved

**Comment by shadowhand** [comment]
:+1:
